internal/framework/status/updater.go

```diff
@@ -213,34 +221,57 @@ func (upd *UpdaterImpl) updateGatewayAPI(ctx context.Context, statuses GatewayAP
 	}
 }
 
+// The function in wait.ExponentialBackoffWithContext will retry if it returns nil as its error,
+// which is what we want if we encounter an error from the functions we call. However,
+// the linter will complain if we return nil if an error was found.
+//
+//nolint:nilerr
 func (upd *UpdaterImpl) writeStatuses(
 	ctx context.Context,
 	nsname types.NamespacedName,
 	obj client.Object,
 	statusSetter func(client.Object),
 ) {
-	// The function handles errors by reporting them in the logs.
-	// We need to get the latest version of the resource.
-	// Otherwise, the Update status API call can fail.
-	// Note: the default client uses a cache for reads, so we're not making an unnecessary API call here.
-	// the default is configurable in the Manager options.
-	if err := upd.cfg.Client.Get(ctx, nsname, obj); err != nil {
-		if !apierrors.IsNotFound(err) {
-			upd.cfg.Logger.Error(
-				err,
-				"Failed to get the recent version the resource when updating status",
-				"namespace", nsname.Namespace,
-				"name", nsname.Name,
-				"kind", obj.GetObjectKind().GroupVersionKind().Kind)
-		}
-		return
-	}
+	// To preserve and log the error message inside the function in wait.ExponentialBackoffWithContext
+	var lastError error
+
+	err := wait.ExponentialBackoffWithContext(
+		ctx,
+		wait.Backoff{
+			Duration: time.Millisecond * 200,
+			Factor:   2,
+			Jitter:   0.5,
+			Steps:    4,
+			Cap:      time.Millisecond * 3000,
+		},
+		// Function returns true if the condition is satisfied, or an error if the loop should be aborted.
+		func(ctx context.Context) (bool, error) {
```

Comment:
I wouldn't say there's a lot of logic that's not covered. It looks like we are just missing a test case for when the object has been deleted. That's going to be hard to test with the current code because `Update` doesn't return an error.

How big of a deal is this? If we really want to test it, I think we'd have to pull out this function and test it separately.

Reply:
it make sense to test it separately. for that, we can mock API client, similarly to how we do that in https://github.com/nginxinc/nginx-gateway-fabric/blob/8e57fe86d311d6a618afa109999d80439d5ca9e9/internal/framework/controller/getter.go#L13

However, it also makes sense in updater test to ensure it have a case when it cannot set the status, meaning `err := wait.ExponentialBackoffWithContext(` for _any_ reason. We can achieve that by creating a client without the scheme (https://github.com/nginxinc/nginx-gateway-fabric/blob/8e57fe86d311d6a618afa109999d80439d5ca9e9/internal/framework/status/updater_test.go#L48-L54)  I don't see any way to confirm that it works other than checking that we logged "Failed to update status", but it is possible though by capturing logs into a buffer.